### ISSUE TITLE
Scylla DB Session: Technical Questions Regarding Sessions

### ISSUE THREAD

**Issue by WarrenN1** [scylladb/scylla-rust-driver#724]
I know you guys are busy and your time is valuable so I will keep this short. My questions are as follows:
1. Does SessionBuilder::new() manage an underlying connection pool or does session builder only have 1 connection?
2. Does SessionBuilder allow for sending multiple queries from multiple threads to the same session which can be completed at different times or does it have a query queue that is blocking in nature ie. sends 1 query, waits for response, sends next query etc.
3. Any recommendations for handling Sessions that makes use of async for proper web server connection pooling.

**Comment by cvybhu** [comment]
Hi,

The general idea is that each application should have a single instance of `Session`. At the beginning you can create a `Session` using `SessionBuilder`, this should be done only once. During `SessionBuilder::build()` the `Session` creates its own pool of connections to the cluster, which is managed automatically. To share a `Session` between many threads it can be wrapped in an `Arc` (`Arc<Session>`). 

> 2. Does SessionBuilder allow for sending multiple queries from multiple threads to the same session which can be completed at different times or does it have a query queue that is blocking in nature ie. sends 1 query, waits for response, sends next query etc.

It allows sending multiple queries from multiple threads. Each call to `query` starts a new request in parallel to the other ones.

To sum up, create a single `Arc<Session>` and share it around your application. Connection pooling and query parallelism is taken care of automatically.

**Comment by WarrenN1** [comment]
Wonderful design. Love to hear it. Wish more libraries/products did this.

**Comment by cvybhu** [comment]
Have fun :)

Closing the issue then.

**Comment by WarrenN1** [comment]
One more quick question, you said to wrap it in an Arc. Does your Session implement Send, Sync in its design or do I need to implement that (via wrapping it in a mutex)?

**Comment by WarrenN1** [comment]
@cvybhu Ok wonderful. Was just checking given the example implied a different underlying structure. Could you point me to any resource in the documentation on the particulars of the implementation/event loop handling?

**Comment by cvybhu** [comment]
> Could you point me to any resource in the documentation on the particulars of the implementation/event loop handling?

There isn't any such documentation. You can try reading the code in `connection.rs` and the adjacent modules.
This is an internal implementation detail that is abstracted away from driver users.
